src/stdlib/mezmo_map_key_values.rs

```diff
@@ -0,0 +1,279 @@
+use crate::compiler::prelude::*;
+
+/// Implements depth first iteration of the keys and values of an object or the values of an array
+///
+/// Key points:
+/// - Recursion depth is controlled via `max_depth`:
+///   - -1: infinite recursion
+///   -  0: no recursion (only current level)
+///   -  N > 0: recurse up to N levels deeper
+/// - Uses Value::into_iter(false) to process items at the current depth
+/// - Recurses into nested objects/arrays up to the specified depth
+/// - Mutates keys/values in-place
+/// - Key collision behavior is based on the sort order of the BTreeMap.
+///   - Value is a BTreeMap; keys are sorted
+///   - On collision, overwrite is based on which key is last in the list
+///   - See lib/tests/tests/mezmo/mezmo_map_key_values/key_collision_overwrite.vrl for an
+///     example of how this may differ from a caller's expectations
+/// - The overhead of this function is the same as using `Value::into_iter(true).by_ref()`
+///   and collecting the final results as used in map_keys and map_values
+fn mezmo_map_key_values<T>(
+    mut value: Value,
+    max_depth: i64,
+    ctx: &mut Context,
+    runner: &closure::Runner<T>,
+) -> Resolved
+where
+    T: Fn(&mut Context) -> Resolved,
+{
+    map_key_values(&mut value, max_depth, ctx, runner)?;
+    Ok(value)
+}
+
+/// Recursively transform `value` in-place up to `remaining_depth`
+fn map_key_values<T>(
+    value: &mut Value,
+    remaining_depth: i64,
+    ctx: &mut Context,
+    runner: &closure::Runner<T>,
+) -> Result<(), ExpressionError>
+where
+    T: Fn(&mut Context) -> Resolved,
+{
+    if !value.is_object() && !value.is_array() {
+        return Ok(());
+    }
+
+    // Take ownership so we can build a non-recursive iterator over the current level.
+    // into_iter requires an owned object
+    let current = core::mem::replace(value, Value::Null);
```

Comment:
It looks like we will leave a `null` in place of the provided top-level `Value` when we encounter an error in any `map_key_values` call while iterating. Might be worth adding a test to demonstrate and confirm that behavior?

If it does, we might need to clone in the caller to provide this function an owned value. Then we can `mem::replace` the original value only if the `map_key_values` call succeeds.